staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -231,90 +268,115 @@ func (c *celAdmissionController) Validate(
 					addConfigError(err, definition, binding)
 
-					if k8serrors.IsNotFound(err) {
-						// Param doesnt exist yet?
-						// Maybe just have to wait a bit.
+					if k8serrors.IsInvalid(err) {
+						// Param mis-configured
+						// require to set paramRef.namespace for namespaced resource and unset paramRef.namespace for cluster scoped resource
 						continue
 					}
 
 					// There was a bad internal error
 					utilruntime.HandleError(err)
 					continue
 				}
+			} else {
+				paramRef := binding.Spec.ParamRef
+				if paramRef != nil {
+					// return error if ValidatingAdmissionPolicyBinding is mis-configured
+					return fmt.Errorf("ValidatingAdmissionPolicyBinding '%s' is not allowed to set paramRef to '%s' since policyName refers to a ValidatingAdmissionPolicy '%s' which has no paramKind",
+						binding.Name, paramRef.String(), definition.Name)
+				}
 			}
 
-			if bindingInfo.evaluator == nil {
+			validator := bindingInfo.validator.Load()
+			if validator == nil {
 				// Compile policy definition using binding
-				bindingInfo.evaluator, err = definition.Compile(c.objectConverter, c.restMapper)
-				if err != nil {
-					// compilation error. Apply failure policy
-					wrappedError := fmt.Errorf("failed to compile CEL expression: %w", err)
-					addConfigError(wrappedError, definition, binding)
-					continue
-				}
-				c.bindingInfos[namespacedBindingName] = bindingInfo
+				newValidator := c.validatorCompiler.Compile(definition)
+				validator = &newValidator
+
+				bindingInfo.validator.Store(validator)
+			}
+
+			decisions, err := (*validator).Validate(a, o, param, matchKind)
+			if err != nil {
+				// runtime error. Apply failure policy
+				wrappedError := fmt.Errorf("failed to evaluate CEL expression: %w", err)
+				addConfigError(wrappedError, definition, binding)
+				continue
 			}
 
-			decisions := bindingInfo.evaluator(a, param)
 			for _, decision := range decisions {
-				switch decision.Kind {
-				case Admit:
-					// Do nothing
-				case Deny:
-					allDecisions = append(allDecisions, PolicyDecisionWithMetadata{
-						Definition:     definition,
-						Binding:        binding,
-						PolicyDecision: decision,
+				switch decision.kind {
+				case admit:
+					// TODO: add metrics for ignored error here
+				case deny:
+					deniedDecisions = append(deniedDecisions, policyDecisionWithMetadata{
+						definition:     definition,
+						binding:        binding,
+						policyDecision: decision,
 					})
 				default:
-					// unrecognized decision. ignore
+					return fmt.Errorf("unrecognized evaluation decision '%s' for ValidatingAdmissionPolicyBinding '%s' with ValidatingAdmissionPolicy '%s'",
+						decision.kind, binding.Name, definition.Name)
 				}
 			}
 		}
 	}
 
-	if len(allDecisions) > 0 {
-		return k8serrors.NewConflict(
-			a.GetResource().GroupResource(), a.GetName(),
-			&PolicyError{
-				Decisions: allDecisions,
-			})
+	if len(deniedDecisions) > 0 {
+		return &policyError{
```

Comment:
This should return a standard error (`api/errors#StatusError`) so it contains the name/kind of the object being rejected and gets displayed the way we want to clients

I'm not sure the `policyError` type is that helpful... it's only being used here, and only when deniedDecisions is of non-zero length, but it is having to guard against zero-length denied decisions, and seems to be duplicating Status error construction already handled in admission.NewForbidden and apierrors.NewForbidden

admission.NewForbidden might be a good starting point for a *StatusError we can tweak the Code/Reason of, like this:

```
// TODO: refactor NewForbidden so the name extraction is reusable for different error types
err := admission.NewForbidden(a, errors.New(message)).(*apierrors.StatusError)
if len(deniedDecisions[0].reason) != 0 {
  err.ErrStatus.Reason = deniedDecisions[0].reason
  err.ErrStatus.Code = reasonToCode(deniedDecisions[0].reason)
}
```

Follow-up: A complication is that kubectl display of errors coming from the server gets *really* twisty depending on their type. Sometimes it displays message, sometimes it displays causes, sometimes it doesn't. Before graduation we need a test that sets up policies and rejected objects with every reason we support, and check what would get returned to the user when running through `k8s.io/kubectl/pkg/cmd/util#CheckErr` to make sure the message from the validation rule is included and not duplicated

Reply:
Code updated. Followup tracked

Reply:
Note to myself: kubectl 1.22 and older might have issue displaying the consistent message. I have tested with earlier version which a apistatus error returned. And with kubectl 1.22 it shows `The request is invalid` without reason/message. Queued up for further look.

Reply:
what 1.22 version did you use?

https://github.com/kubernetes/kubernetes/pull/112150 and https://github.com/kubernetes/kubernetes/issues/111108 might be related